kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,623 @@
+     * @param rValue            Value to be used in assigning
+     */
+    void Assign(
+        const GlobalPointer<TPointerDataType>& rGlobalPointer,
+        const TValueDataType& rValue)
+    {
+        (this->*(this->mUpdateMethod))(rGlobalPointer, rValue);
+    }
+
+    /**
+     * @brief This method does all the communication
+     *
+     * This method should be called once (or least amount of times) since this involves doing
+     * mpi communication to update remote gp values.
+     *
+     * This does not have any cost in the serial run.
+     *
+     * Ghost mesh synchronization needs to be done afterwards.
+     */
+    void SendAndApplyRemotely()
+    {
+        mrPointerCommunicator.SendAndApplyRemotely(*this);
+    }
+
+    ///@}
+
+private:
+    ///@name Private members
+    ///@{
+
+    const int mCurrentRank;
+    const TApplyFunctor& mrApplyFunctor;
+
+    void (ProxyType::*mUpdateMethod)(const GlobalPointer<TPointerDataType>&, const TValueDataType&);
+
+    // here we store this in two vector maps(vector of gps, vector of values) instead of one vector map(key:gp, value: std::vector)
+    // because, when updating
+    //      1. Two vector maps will have to only have to append the two vectors for each non-local update.
+    //          Whenever std::vector reaches the capacity, it is doubled,
+    //          so the cost of resizing for vector appending is in O(log N)
+    //      2. If a single vector map is used, first it has to find the gp in the map [ O(log N) if found, otherwise larger] and then append one vector
+    //          for each non-local update.
+    // The method 2 is little bit more expensive than first one AFAIK. so implemented the vector versions. Suggestions are highly appreciated :)
+    //
+    std::vector<GlobalPointerMapType> mNonLocalGlobalPointerMapsVector;
+    std::vector<DataVectorMapType> mNonLocalDataValueMapsVector;
+
+    TGPMapCommunicator& mrPointerCommunicator;
+
+    ///@}
+    ///@name Private operations
+    ///@{
+
+    /**
+     * @brief Assign local values
+     *
+     * This method assigns local gp values using mrApplyFunctor.
+     * It assumes rGlobalPointer is a local gp always, therefore
+     * no checks are performed.
+     *
+     * @param rGlobalPointer    Local Global pointer of the destination
+     * @param rValue            Value to be used in assigning
+     */
+    void AssignLocalData(
+        const GlobalPointer<TPointerDataType>& rGlobalPointer,
+        const TValueDataType& rValue)
+    {
+        KRATOS_TRY
+
+        KRATOS_DEBUG_ERROR_IF(rGlobalPointer.GetRank() != mCurrentRank)
+            << "Using local global pointer update method with a non-local "
+               "global pointer. [ MyPID = "
+            << mCurrentRank
+            << " GlobalPointerRank = " << rGlobalPointer.GetRank() << " ].\n";
+
+        auto gp_pointer = rGlobalPointer;  // required to remove the const from rGlobalPointer
```

Comment:
this is a hack.  i think in this case you need to take out the const on line 270, or even there were u make  the loop

Reply:
If we remove the const in line 270, then There is a problem. In the following block
https://github.com/KratosMultiphysics/Kratos/blob/c0d4c866ead2d10084a336b096b61577abaf800f/kratos/utilities/pointer_map_communicator.h#L452-L465

We will need to do this hack before line 463 because `std::unordered_map` always store keys with the qualifier `const`. In there we will need to remove this `const` qualifier before passing it to the proxy.

I changed the implementation to use the `move` constructor which is efficient. Would that be ok?